### PR TITLE
fix json unmarshal bug when extension used with Responses

### PR DESCRIPTION
**Pull request by alexzielenski**
discovered while implementing https://github.com/kubernetes/kube-openapi/pull/319

spec allows vendor extensions, but if they are used the parsing would fail and leave both `Default` and `StatusCodeResponses` empty

code now decodes each response individually so vendor extensions can be filtered out

### CODE REVIEW

File: pkg/validation/spec/responses.go [hunk 1]
```diff
@@ -90,20 +90,28 @@ func (r ResponsesProps) MarshalJSON() ([]byte, error) {
 
 // UnmarshalJSON unmarshals responses from JSON
 func (r *ResponsesProps) UnmarshalJSON(data []byte) error {
-	var res map[string]Response
+	var res map[string]json.RawMessage
 	if err := json.Unmarshal(data, &res); err != nil {
 		return nil
```
Comment (apelisse) [inline pkg/validation/spec/responses.go:95]: why do we return nil here?

Comment (apelisse) [reply pkg/validation/spec/responses.go:95]: Also, I've tried to switch that to `return err`, but it still doesn't fail even when it should. Are there other occurences of errors not being propagated properly?

**Comment by apelisse** [comment]
/lgtm
/approve

Looks good, I left one question but it's only slightly related.

File: pkg/validation/spec/responses.go (hunk 1 above)
Comment (alexzielenski) [reply pkg/validation/spec/responses.go:95]: > but it still doesn't fail even when it should. 

I'm not sure what you've tried, but it should fail if provided with a non-map element i.e. `["a", "b", "c"]`, or `123`.

you are right that this should not return nil

File: pkg/validation/spec/responses.go [hunk 2]
```diff
@@ -90,20 +90,28 @@ func (r ResponsesProps) MarshalJSON() ([]byte, error) {
...
-	var res map[string]Response
+	var res map[string]json.RawMessage
 	if err := json.Unmarshal(data, &res); err != nil {
-		return nil
+		return err
 	}
 	if v, ok := res["default"]; ok {
-		r.Default = &v
+		value := Response{}
+		json.Unmarshal(v, &value)
```
Comment (liggitt) [inline pkg/validation/spec/responses.go:99]: don't we need to check for `err` here and return if this errors?

Comment (alexzielenski) [reply pkg/validation/spec/responses.go]: whoops. thanks for catching that

File: pkg/validation/spec/responses.go [hunk 3]
```diff
@@ -90,20 +90,30 @@ func (r ResponsesProps) MarshalJSON() ([]byte, error) {
...
 	}
 	for k, v := range res {
+		// Take all integral keys
 		if nk, err := strconv.Atoi(k); err == nil {
 			if r.StatusCodeResponses == nil {
 				r.StatusCodeResponses = map[int]Response{}
 			}
-			r.StatusCodeResponses[nk] = v
+			value := Response{}
+			err := json.Unmarshal(v, &value)
```
Comment (apelisse) [inline pkg/validation/spec/responses.go:112]: Why not consistent?!

**Pull request merged**